libraries/AP_BattMonitor/AP_BattMonitor_MAVLink_Battery.cpp

```diff
@@ -0,0 +1,129 @@
+#include <AP_HAL/AP_HAL.h>
+#include <GCS_MAVLink/GCS_MAVLink.h>
+#include "AP_BattMonitor_MAVLink_Battery.h"
+
+AP_BattMonitor_MAVLink_Battery::AP_BattMonitor_MAVLink_Battery(AP_BattMonitor &mon,
+                                             AP_BattMonitor::BattMonitor_State &mon_state,
+                                             AP_BattMonitor_Params &params) :
+    AP_BattMonitor_Backend(mon, mon_state, params)
+{
+	_state.healthy = false;
+}
+
+void AP_BattMonitor_MAVLink_Battery::read()
+{
+    if ((AP_HAL::micros() - _state.last_time_micros) > MAVLINK_SMART_BATTERY_TIMEOUT_MICROS) {
+		_state.healthy = false;
+		return;
+	} 
+	_state.healthy = true;	
+   return;
+}
+
+bool AP_BattMonitor_MAVLink_Battery::capacity_remaining_pct(uint8_t &cap_remain_pct) const
+{
+    if (capacity_rem_pct != INT8_MIN) {
+		/* positive capacity remaining is expected */
+		if (capacity_rem_pct > 0) {
+			cap_remain_pct = capacity_rem_pct;
+		} else {
+			cap_remain_pct = 0;
+		}
+    	return true;
+	}
+	return false;
+}
+
+bool AP_BattMonitor_MAVLink_Battery::get_cycle_count(uint16_t &cycles) const
+{
+    if (cycle_count != UINT16_MAX) {
+		cycles = cycle_count;
+    	return true;
+	}
+	return false;
+}
+
+void AP_BattMonitor_MAVLink_Battery::process_info_msg(const mavlink_message_t &msg)
+{
+	int32_t capacity = mavlink_msg_smart_battery_info_get_capacity_full(&msg);
+	if (capacity != -1
+		&& capacity != _params._pack_capacity) {
+		_params._pack_capacity.set_and_notify(capacity);
+	}
+	cycle_count = mavlink_msg_smart_battery_info_get_cycle_count(&msg);
+}
+
+void AP_BattMonitor_MAVLink_Battery::process_status_msg(const mavlink_message_t &msg)
+{
+	__mavlink_battery_status_t bat_stat;
+	mavlink_msg_battery_status_decode(&msg, &bat_stat);
+	fault_bitmask = bat_stat.fault_bitmask;

```

Comment:
make sure you do an ID check. If you have two batteries, how do you know which one talks to which driver? With the BATTn_SERIAL_NUM param! Take a look at AP_BattMonitor_UAVCAN::match_battery_id()